### PR TITLE
[#3891] feat(catalog-lakehouse-paimon): Support schema operations for Paimon Catalog

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?
Support schema operations for Paimon Catalog.

### Why are the changes needed?

Fix: https://github.com/datastrato/gravitino/issues/3891

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New Uts and ITs.

### CODE REVIEW

**Comment by caican00** [comment]
Hi @FANNG1 could you help review this PR? Thanks!

**Comment by caican00** [comment]
> Hi @FANNG1 could you help review this PR? Thanks!

could you help re-trigger the pipeline? Thanks

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 1]
```diff
@@ -98,7 +165,7 @@ public PaimonSchema loadSchema(NameIdentifier identifier) throws NoSuchSchemaExc
   @Override
   public PaimonSchema alterSchema(NameIdentifier identifier, SchemaChange... changes)
       throws NoSuchSchemaException {
-    throw new UnsupportedOperationException();
+    throw new UnsupportedOperationException("Alter schema is not supported in Paimon Catalog.");
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:168]: Paimon doesn't support alter schema or we doesn't support ?

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 2]
```diff
@@ -206,5 +284,17 @@ public boolean purgeTable(NameIdentifier identifier) throws UnsupportedOperation
...
+    if (paimonTableOps != null) {
+      try {
+        paimonTableOps.close();
+      } catch (Exception e) {
+        throw new IOException(e.getMessage());
+      }
+    }
+  }
+
+  private static String currentUser() {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:297]: seems should use  `UserGroupInformation.getCurrentUser().getUserName()` to get current user, @qqqttt123 @yuqi1129 help help to check it

Comment (yuqi1129) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:297]: We should use `PrincipalUtils.getCurrentUserName` to get the current user name.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonSchema.java [hunk 3]
```diff
@@ -4,9 +4,14 @@
...
 package com.datastrato.gravitino.catalog.lakehouse.paimon;
 
+import static com.datastrato.gravitino.meta.AuditInfo.EMPTY;
+
 import com.datastrato.gravitino.Schema;
 import com.datastrato.gravitino.connector.BaseSchema;
+import java.util.Map;
+import java.util.Optional;
 import lombok.ToString;
+import org.apache.paimon.utils.Pair;
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonSchema.java:14]: use `org.apache.commons.lang3.tuple.Pair` ?

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/ops/PaimonTableOps.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.List;
+import java.util.Map;
+import org.apache.paimon.catalog.Catalog;
+import org.apache.paimon.catalog.Catalog.DatabaseAlreadyExistException;
+import org.apache.paimon.catalog.Catalog.DatabaseNotEmptyException;
+import org.apache.paimon.catalog.Catalog.DatabaseNotExistException;
+import org.apache.paimon.utils.Pair;
+
+/** Table operation proxy that handles table operations of an underlying Paimon catalog. */
+public class PaimonTableOps implements AutoCloseable {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/ops/PaimonTableOps.java:19]: PaimonCatalogOps?

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 1 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:168]: paimon

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 2 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:297]: got it.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonSchema.java (hunk 3 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonSchema.java:14]: ok for me.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/ops/PaimonTableOps.java (hunk 4 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/ops/PaimonTableOps.java:19]: ok.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+   * @return The {@link Catalog} instance of catalog backend.
+   */
+  public static Catalog loadCatalogBackend(PaimonConfig paimonConfig) {
+    String metastore = paimonConfig.get(CATALOG_BACKEND);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(metastore), "Paimon Catalog metastore can not be null or empty.");
+    String warehouse = paimonConfig.get(CATALOG_WAREHOUSE);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(warehouse), "Paimon Catalog warehouse can not be null or empty.");
+    if (!PaimonCatalogBackend.FILESYSTEM
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java:39]: equalIgnoreCase

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java:39]: got it.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 1 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:168]: https://github.com/apache/paimon/blob/master/paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java

**Comment by caican00** [comment]
comments have been addressed. please help review it again, Thank you! cc @FANNG1 @yuqi1129

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 6]
```diff
@@ -206,5 +285,17 @@ public boolean purgeTable(NameIdentifier identifier) throws UnsupportedOperation
   }
 
   @Override
-  public void close() throws IOException {}
+  public void close() throws IOException {
+    if (paimonCatalogOps != null) {
+      try {
+        paimonCatalogOps.close();
+      } catch (Exception e) {
+        throw new IOException(e.getMessage());
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:293]: why throwing `IOException`?

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogPropertiesMetadata.java [hunk 7]
```diff
@@ -4,21 +4,66 @@
...
+import java.util.List;
 import java.util.Map;
 
 /**
  * Implementation of {@link PropertiesMetadata} that represents Paimon catalog properties metadata.
  */
 public class PaimonCatalogPropertiesMetadata extends BaseCatalogPropertiesMetadata {
 
-  private static final Map<String, PropertyEntry<?>> PROPERTIES_METADATA = ImmutableMap.of();
+  public static final String GRAVITINO_CATALOG_BACKEND = "catalog-backend";
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogPropertiesMetadata.java:26]: please check whether the variable and method should be `public`

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    String metastore = paimonConfig.get(CATALOG_BACKEND);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(metastore), "Paimon Catalog metastore can not be null or empty.");
+    String warehouse = paimonConfig.get(CATALOG_WAREHOUSE);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(warehouse), "Paimon Catalog warehouse can not be null or empty.");
+    if (!PaimonCatalogBackend.FILESYSTEM.name().equalsIgnoreCase(metastore)) {
+      String uri = paimonConfig.get(CATALOG_URI);
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(uri), "Paimon Catalog uri can not be null or empty.");
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java:41]: please add specific metastore to the error message

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/integration/test/CatalogPaimonBaseIT.java [hunk 9]
```diff
@@ -57,7 +72,75 @@ private void resetSchema() {
 
   protected abstract Map<String, String> initPaimonCatalogProperties();
 
-  private void clearTableAndSchema() {}
+  @Test
+  void testPaimonSchemaOperations() throws DatabaseNotExistException {
+    SupportsSchemas schemas = catalog.asSchemas();
+
+    // list schema check.
+    Set<String> schemaNames = new HashSet<>(Arrays.asList(schemas.listSchemas()));
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/integration/test/CatalogPaimonBaseIT.java:80]: please don't rely on `schemaName`, it's out of the test

Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/integration/test/CatalogPaimonBaseIT.java:80]: it's odd to test list schema first

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+            .withName(name)
+            .withNamespace(Namespace.of(META_LAKE_NAME))
+            .withType(PaimonCatalog.Type.RELATIONAL)
+            .withProvider("lakehouse-paimon")
+            .withAuditInfo(AUDIT_INFO)
+            .build();
+
+    Map<String, String> conf = Maps.newHashMap();
+    conf.put(PaimonCatalogPropertiesMetadata.GRAVITINO_CATALOG_BACKEND, "filesystem");
+    conf.put(PaimonCatalogPropertiesMetadata.WAREHOUSE, "/tmp/paimon_catalog_warehouse");
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java:160]: I'm afraid the ut may failed on windows for `/tmp/paimon_catalog_warehouse`

Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java:160]: could the ut works if running multi times? there are some state saved in filesystem

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        @Override
+        public PropertiesMetadata topicPropertiesMetadata() throws UnsupportedOperationException {
+          throw new UnsupportedOperationException("Topic properties are not supported");
+        }
+      };
+
+  @Test
+  public void testListDatabases() {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java:57]: why putting `testListDatabases` here?

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java (hunk 10 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java:160]: yes, there is a cleanup operation.

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java (hunk 11 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java:57]: just to make sure the catalog object works.

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java (hunk 10 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonSchema.java:160]: ok, i will fix this.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 12]
```diff
@@ -71,7 +110,29 @@ public NameIdentifier[] listSchemas(Namespace namespace) throws NoSuchCatalogExc
...
+    PaimonSchema createdSchema =
+        PaimonSchema.builder()
+            .withName(identifier.name())
+            .withComment(comment)
+            .withProperties(properties)
+            .withAuditInfo(
+                AuditInfo.builder().withCreator(currentUser).withCreateTime(Instant.now()).build())
+            .build();
+    try {
+      paimonCatalogOps.createDatabase(createdSchema.toPaimonSchema());
```
[the remark this answers is not in the document]
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:123]: How about `toOriginalPaimonSchema` and `fromOriginalPaimonSchema` ? @FANNG1

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 13]
```diff
@@ -71,7 +110,29 @@ public NameIdentifier[] listSchemas(Namespace namespace) throws NoSuchCatalogExc
...
+    } catch (Catalog.DatabaseAlreadyExistException e) {
+      throw new SchemaAlreadyExistsException(e, SCHEMA_ALREADY_EXISTS_EXCEPTION, identifier);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    LOG.info(
+        "Created Paimon schema (database): {}.\nCurrent user: {} \nComment: {}.\nMetadata: {}.",
```
[the remark this answers is not in the document]
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:130]: update.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 6 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:293]: fixed.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogPropertiesMetadata.java (hunk 7 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogPropertiesMetadata.java:26]: updated.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java (hunk 8 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/utils/CatalogUtils.java:41]: updated.

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/integration/test/CatalogPaimonBaseIT.java (hunk 9 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/integration/test/CatalogPaimonBaseIT.java:80]: updated.

**Comment by caican00** [comment]
@FANNG1 could you help review again? Thank you!

**Comment by caican00** [comment]
Hi @FANNG1 @yuqi1129 could you help review again if you are free? Thanks!

File: catalogs/catalog-lakehouse-paimon/build.gradle.kts [hunk 14]
```diff
@@ -41,6 +42,8 @@ dependencies {
   testImplementation(project(":integration-test-common", "testArtifacts"))
   testImplementation(project(":server"))
   testImplementation(project(":server-common"))
+  testImplementation(libs.bundles.log4j)
+  testImplementation(libs.slf4j.api)
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-paimon/build.gradle.kts:46]: Please keep the alphabetical order

**Comment by yuqi1129** [comment]
I have no further comments except the one above.

File: catalogs/catalog-lakehouse-paimon/build.gradle.kts (hunk 14 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/build.gradle.kts:44]: updated. cc @yuqi1129

**Review by yuqi1129 — DISMISSED** [review]
LGTM

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java (hunk 11 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java:61]: maybe you should use a better name

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java]: How about `toPaimonProperties` and `fromPaimonProperties`?

**Comment by FANNG1** [comment]
LGTM, except some naming

File: catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java (hunk 11 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/com/datastrato/gravitino/catalog/lakehouse/paimon/TestPaimonCatalog.java]: updated.

File: catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java]: updated.

**Comment by caican00** [comment]
comments have been addressed. cc @FANNG1

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00  merged to main,  thanks for your work

**Pull request merged**